src/app/tests/TestDataModelSerialization.cpp

```diff
@@ -0,0 +1,833 @@
+
+            // Encode every field manually except a.
+            {
+                err = EncodeStruct(_this->mWriter, TLV::AnonymousTag,
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kAFieldId), t.a),
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kBFieldId), t.b),
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kCFieldId), t.c),
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kDFieldId), t.e),
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kEFieldId), t.e));
+                NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+            }
+
+            err = _this->mWriter.Finalize();
+            NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+            _this->DumpBuf();
+        }
+
+        //
+        // Decode
+        //
+        {
+            clusters::TestCluster::SimpleStruct::Type t;
+
+            _this->SetupReader();
+
+            err = DataModel::Decode(_this->mReader, t);
+            NL_TEST_ASSERT(apSuite, err != CHIP_NO_ERROR);
+        }
+    }
+}
+
+void TestDataModelSerialization::TestDataModelSerialization_InvalidListType(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err;
+    auto * _this = static_cast<TestDataModelSerialization *>(apContext);
+
+    _this->mpSuite = apSuite;
+    _this->SetupBuf();
+
+    //
+    // Encode
+    //
+    {
+        clusters::TestCluster::NestedStructList::Type t;
+        uint32_t intBuf[4] = { 10000, 10001, 10002, 10003 };
+
+        t.e = intBuf;
+
+        // Encode a list of integers for field d instead of a list of structs.
+        {
+            err = EncodeStruct(_this->mWriter, TLV::AnonymousTag,
+                               MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::NestedStructList::kDFieldId), t.e));
+            NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+        }
+
+        err = _this->mWriter.Finalize();
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        _this->DumpBuf();
+    }
+
+    //
+    // Decode
+    //
+    {
+        clusters::TestCluster::DecodableNestedStructList::Type t;
+
+        _this->SetupReader();
+
+        err = DataModel::Decode(_this->mReader, t);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        auto iter     = t.d.begin();
+        bool hadItems = false;
+
+        while (iter.Next())
+        {
+            hadItems = true;
+        }
+
+        NL_TEST_ASSERT(apSuite, iter.GetStatus() != CHIP_NO_ERROR);
+        NL_TEST_ASSERT(apSuite, !hadItems);
+    }
+}
+
+} // namespace app
+} // namespace chip
+
+namespace {
+
+void InitializeChip(nlTestSuite * apSuite)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = chip::Platform::MemoryInit();
```

Comment:
This needs a corresponding shutdown, right?  Should have caught this the first time....

Probably should use the actual init/shutdown functions from the `nlTestSuite` struct for this, like other tests do.